### ISSUE TITLE
JUnit5 - AfterEach callback is leaking connection

### ISSUE THREAD

**Issue by rmpestano** [database-rider/database-rider#346]
While fixing #331  we found a leaking connection issue and it turns out that when using `cacheConnection=false` and having a rider annotation in after each (`@DataSet` or `@ExpectedDataSet`) will make a connection to be left open:

```
        @AfterEach
	@ExpectedDataSet("expected.yml")
	void afterEach() {
		dataSource2 = dataSource;
	} 
	
	@AfterAll
	static void afterAll() {
		WrappingDataSource wrappingDataSource = (WrappingDataSource) dataSource2;
		assertEquals(0, wrappingDataSource.liveConnections.get());
	}
```

**Comment by hheg** [comment]
>EDIT2: By the way checking 0 living connections inside the test (assertEquals(0, wrappingDataSource.liveConnections.get());) will never be true because we close it after test execution

If there's no cached connection there shouldn't be any open connections at this stage, since assume that you have an `@AfterEach` declared method there's no guarantee that there's a `@AfterAll` method to close, so each method that opens a non cached connection must be responsible for closing it otherwise it will leak.

**Comment by rmpestano** [comment]
> so each method that opens a non cached connection must be responsible for closing it otherwise it will leak.

That's really not efficient, what we do instead is get a connection in the test (or in a before callback) and reuse it for seeding (`@DataSet`), cleaning (`cleanBefore/CleanAfter`), disablingContraints and for comparing the DB after the test (`@ExpectedDataSet`) and then after test execution we close it. I'm considering closing it in the after callback but needs some refactoring

**Comment by hheg** [comment]
> > so each method that opens a non cached connection must be responsible for closing it otherwise it will leak.
> 
> That's really not efficient, what we do instead is get a connection in the test (or in a before callback) and reuse it for seeding (`@DataSet`), cleaning (`cleanBefore/CleanAfter`), disablingContraints and for comparing the DB after the test (`@ExpectedDataSet`) and then after test execution we close it. I'm considering closing it in the after callback but needs some refactoring

But the problem is that you don't know if there will be a `@AfterEach/All` method declared in the test. Also if you cache/store these connections that wouldn't that mean that the cacheConnection=false flag will be useless? Like if I use Hikari's own connection leak detection mechanics, that connection will start showing up as a leaking connection in each test and pollute our hunt for actual leaking connections. I don't think a test framework should infer such side effects when it's job is just test the code.

How much less efficient is it to close the connection every time during each callback? Especially if you use a connection pool the connection will never be closed, just returned to the pool.

**Comment by rmpestano** [comment]
> But the problem is that you don't know if there will be a @AfterEach/All method declared in the test.

JUnit5 engine will always run the [extension after each/all callbacks](https://github.com/database-rider/database-rider/blob/da32cce800c1a8b03ebccdc861d5bc843367152b/rider-junit5/src/main/java/com/github/database/rider/junit5/DBUnitExtension.java#L155)

> Also if you cache/store these connections that wouldn't that mean that the cacheConnection=false flag will be useless?

There will be a connection per test method while `cacheConnection=true` the same connection is used per test class

> How much less efficient is it to close the connection every time during each callback? Especially if you use a connection pool the connection will never be closed, just returned to the pool.

I did a video a while ago (2016) which is not a real benchmark but without a pool of connections and just a few tests the improvement was already noticeable: https://www.youtube.com/watch?v=uwp9hENKeeI
 
I really don't see why using `cacheConnection=true` is a problem, I think we should always cache the connection on the tests but since we have this option I'm whiling to fix the issues. Maybe in rider 2.0 we might remove cacheConnection=false

**Comment by hheg** [comment]
> JUnit5 engine will always run the extension after each/all callbacks
Then it's not that far of using and closing a connection per callback, instead of holding on to it during three?

>I really don't see why using cacheConnection=true is a problem, I think we should always cache the connection on the tests because tests are short-lived workloads. But since we have this option I'm willing to fix the issues. Maybe in rider 2.0 we'll remove cacheConnection=false

The cached connection will be considered leaked by Hikari's connection leak detection and show up as a false positive which pollutes actual detection since you always needs to disregard one. 

I don't think one can assume that tests are short lived, a test can do a lot of things depending on what it tries to do. It could be used for more complex IT tests and setting those up sometimes takes a while particularly have many other components.

>I recorded a video a while ago (2016) which is not a real benchmark but without a pool of connections and just a few tests the improvement was already noticeable: https://www.youtube.com/watch?v=uwp9hENKeeI

That's fine, but you usually use a pool if you are using an application since using a pool is more efficient and removes the need of holding on to it and manually managing connections are usually error prone. A connection should be short lived, meaning it should be closed as soon as possible after the code is done with it. Dbrider seems to follow the datasource pattern, and the datasource is a connection pool or just holding onto a single "cached" connection is an implementation detail. The code could just close the connection and then, if needed, it could be managed behind the scenes depending on what implementation it is. Calling close on it would then fit most connection leak detection tools expectations.

**Comment by rmpestano** [comment]
> The cached connection will be considered leaked by Hikari's connection leak detection and show up as a false positive which pollutes actual detection since you always needs to disregard one.

One could disable that checking on tests but anyway, I'm accepting PRs if you think that can be improved.

Regarding this issue,  can you check it again? the leak should be gone, just make sure to do the assertion after test execution:

```
        @AfterEach
	@ExpectedDataSet("expected.yml")
	void afterEach() {
		dataSource2 = dataSource;
        WrappingDataSource wrappingDataSource = (WrappingDataSource) dataSource;
        assertEquals(0, wrappingDataSource.liveConnections.get());
	}
	
	@AfterAll
	static void afterAll() {
		WrappingDataSource wrappingDataSource = (WrappingDataSource) dataSource2;
		assertEquals(0, wrappingDataSource.liveConnections.get());
	}

```

**Comment by rmpestano** [comment]
The fix was release in v`1.27.0`, if the issue persists please open a new issue with the details, thanks!
